cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/transforms/StructuredRecordToCubeFactTransform.java

```diff
@@ -0,0 +1,388 @@
+    private final TimestampResolver timestampResolver;
+    private final Collection<TagValueResolver> tagResolvers;
+    private final Collection<MeasurementResolver> measurementResolvers;
+
+    public CubeFactBuilder(MappingConfig config) {
+      this.timestampResolver = new TimestampResolver(config.timestamp);
+      if (config.tags == null) {
+        throw new IllegalArgumentException("Missing required 'tags' configuration: " + config);
+      }
+      this.tagResolvers = Lists.newArrayList();
+      for (ValueMapping mapping : config.tags) {
+        tagResolvers.add(new TagValueResolver(mapping));
+      }
+      if (config.measurements == null) {
+        throw new IllegalArgumentException("Missing required 'measurements' configuration: " + config);
+      }
+      this.measurementResolvers = Lists.newArrayList();
+      for (ValueMapping mapping : config.measurements) {
+        measurementResolvers.add(new MeasurementResolver(mapping));
+      }
+    }
+
+    public CubeFact build(StructuredRecord record) {
+      CubeFact fact = new CubeFact(timestampResolver.getTimestamp(record));
+      for (TagValueResolver resolver : tagResolvers) {
+        String value = resolver.getValue(record);
+        if (value != null) {
+          fact.addTag(resolver.getName(), value);
+        }
+      }
+      for (MeasurementResolver resolver : measurementResolvers) {
+        Measurement measurement = resolver.getMeasurement(record);
+        if (measurement != null) {
+          fact.addMeasurement(measurement);
+        }
+      }
+      return fact;
+    }
+  }
+
+  private static final class TimestampResolver {
+    private String srcField;
+    private DateFormat dateFormat;
+
+    public TimestampResolver(ValueMapping valueMapping) {
+      if (valueMapping == null) {
+        throw new IllegalArgumentException("Missing required configuration for CubeFact timestamp.");
+      }
+      // Timestamp is either retrieved from the record field or current ts (at processing) is used.
+      // If record field is used, either dateFormat is used to parse the value, or it is assumed as epoch in ms
+      if (valueMapping.sourceField != null) {
+        this.srcField = valueMapping.sourceField;
+        if (valueMapping.sourceFieldFormat != null) {
+          this.dateFormat = new SimpleDateFormat(valueMapping.sourceFieldFormat);
+        }
+      } else if (!"now".equals(valueMapping.value)) {
+        throw new IllegalArgumentException("Invalid configuration for CubeFact timestamp: " + valueMapping);
+      }
+    }
+
+    public long getTimestamp(StructuredRecord record) {
+      if (srcField == null) {
+        return System.currentTimeMillis();
+      }
+      Object val = record.get(srcField);
+      if (val == null) {
+        throw new IllegalArgumentException("Required field to determine timestamp is missing: " + srcField);
+      }
+      String valAsString = val.toString();
+      if (dateFormat != null) {
```

Comment:
may be better to require a timestamp, and have a separate transform that parses dates